### PR TITLE
P20-813: Lock account linking until parent permission is granted

### PR DESCRIPTION
**Pull request by artem-vavilov**
## Screenshot
locked-account-linking

## Links
- JIRA ticket: [P20-813](https://codedotorg.atlassian.net/browse/P20-813)
- JIRA ticket: [P20-944](https://codedotorg.atlassian.net/browse/P20-944)
- Figma: [Account Details Locked](https://www.figma.com/file/wR0Myx8T2dHMaeeb1GAabb/CPA-Flows?type=design&node-id=950-863&mode=design&t=zdTDIzqs2lBe9lsf-4)

### CODE REVIEW

File: dashboard/test/controllers/omniauth_callbacks_controller_test.rb [hunk 1]
```diff
@@ -1671,91 +1671,109 @@ class OmniauthCallbacksControllerTest < ActionController::TestCase
...
+          end
+        end
+      end
+    end
   end
 
-  test 'Microsoft SSO: links an LTI auth option to an existing account' do
-    DCDO.stubs(:get).with('lti_account_linking_enabled', false).returns(true)
-    lti_integration = create :lti_integration
+  describe '#link_accounts' do
```
Comment (carl-codeorg) [inline dashboard/test/controllers/omniauth_callbacks_controller_test.rb:1707]: Nice, this really cleans up these tests.

File: dashboard/app/controllers/omniauth_callbacks_controller.rb [hunk 2]
```diff
@@ -527,6 +529,16 @@ def login
...
   end
 
+  private def check_account_linking_lock
+    user = current_user || find_user_by_credential
+    return unless user
+
+    lock_reason = account_linking_lock_reason(user)
+    return unless lock_reason
+
+    redirect_to edit_user_registration_path, alert: lock_reason
```
Comment (carl-codeorg) [inline dashboard/app/controllers/omniauth_callbacks_controller.rb:539]: If the user isn't logged in will this redirect them to the sign-in page? Just need to make sure that it works for users who haven't finished registration yet.

Comment (artem-vavilov) [reply dashboard/app/controllers/omniauth_callbacks_controller.rb:539]: I implemented behavior similar to `OmniauthCallbacksController#connect_provider`. What should we do for `#link_account` then?
https://github.com/code-dot-org/code-dot-org/blob/abcfd4d83a47adbbaac93c787454243255ca0b07/dashboard/app/controllers/omniauth_callbacks_controller.rb#L88-L91

Comment (carl-codeorg) [reply dashboard/app/controllers/omniauth_callbacks_controller.rb:539]: I think it's fine to either redirect them to the page they came from, or to the sign-in page, as long as the alert is present.

**Comment by artem-vavilov** [comment]
This has already been implemented in https://github.com/code-dot-org/code-dot-org/pull/58343

File: dashboard/app/controllers/omniauth_callbacks_controller.rb [hunk 3]
```diff
@@ -11,6 +11,8 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
   # TODO: figure out how to avoid skipping CSRF verification for Powerschool
   skip_before_action :verify_authenticity_token, only: :powerschool
 
+  before_action :check_account_linking_lock, on: %i[connect_provider link_accounts]
```
Comment (wilkie) [inline dashboard/app/controllers/omniauth_callbacks_controller.rb:14]: I missed this again when I cherry-picked your changes, here. The `on` needs to be `only`. Why does it let you even do that!! 😢

**Pull request closed without merging**